airbyte-ci/connectors/pipelines/pipelines/models/steps.py

```diff
@@ -155,14 +158,49 @@ class Step(ABC):
     # The max duration of a step run. If the step run for more than this duration it will be considered as timed out.
     # The default of 5 hours is arbitrary and can be changed if needed.
     max_duration: ClassVar[timedelta] = timedelta(hours=5)
-
     retry_delay = timedelta(seconds=10)
+    accept_extra_params: bool = False
 
     def __init__(self, context: PipelineContext) -> None:  # noqa D107
         self.context = context
         self.retry_count = 0
         self.started_at: Optional[datetime] = None
         self.stopped_at: Optional[datetime] = None
+        self._extra_params: STEP_PARAMS = {}
+
+    @property
+    def extra_params(self) -> STEP_PARAMS:
+        return self._extra_params
+
+    @extra_params.setter
+    def extra_params(self, value: STEP_PARAMS) -> None:
+        if not self.accept_extra_params:
+            raise ValueError(f"{self.__class__.__name__} does not accept extra params.")
+        self._extra_params = value
+
+    @property
+    def default_params(self) -> STEP_PARAMS:
+        return {}
+
+    @property
+    def params(self) -> STEP_PARAMS:
```

Comment:
❓Ok so big question which can change the implementation.

Given that pipelines currently have shown a need to pass arguments between steps
```py
steps: STEP_TREE = [
        [StepToRun(id=CONNECTOR_TEST_STEP_ID.BUILD_TAR, step=BuildConnectorDistributionTar(context))],
        [StepToRun(id=CONNECTOR_TEST_STEP_ID.UNIT, step=UnitTests(context), depends_on=[CONNECTOR_TEST_STEP_ID.BUILD_TAR])],
        [
            StepToRun(
                id=CONNECTOR_TEST_STEP_ID.BUILD,
                step=BuildConnectorImages(context),
                args=lambda results: {
                    "dist_dir": results[CONNECTOR_TEST_STEP_ID.BUILD_TAR].output_artifact.directory(dist_tar_directory_path(context))
                },
                depends_on=[CONNECTOR_TEST_STEP_ID.BUILD_TAR],
            ),
        ],
    ]
    ```
https://github.com/airbytehq/airbyte/blob/773d3bfc027ba7d21488fc3093cf5912ba32231b/airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/steps/java_connectors.py#L155

Can we forsee them also wanting to alter extra args?

Or can we forsee Generic steps like a `PytestStep` preferring devs to use extra_args as the main interface to set additional options?

Reply:
> Given that pipelines currently have shown a need to pass arguments between steps
> Can we forsee them also wanting to alter extra args?

My stand on it is the following:
* Our `connector test` command runs "hardcoded" DAG of steps whose order and dependency should not be altered by user inputs. I consider inter step arg passing being a kind of dependency expression and don't want dev to be able to alter this.

IMO extra parameters should only be options passed to test steps which ideally should not modify the content of artifacts passed in between steps.

This is why I think extra params should not change the `_run` signature.

Reply:
Fair enough! Convinced